lib/vdc/LayeredGrid.cpp

```diff
@@ -129,98 +123,148 @@ void LayeredGrid::GetBoundingBox(
+	bool found = _sg2d.GetIndicesCell(coords, indices, wgts);
+	if (! found) return(found);
 
-	size_t indices1[3];
-	for (int i=0; i<3; i++) {
-		indices1[i] = indices0[i] + 1;
-	}
 
+	// XZ and YZ cell sides are planar, but XY sides may not be. We divide
+	// the XY faces into two triangles (changing hexahedrals into prims)
+	// and figure out which triangle (prism) the point is in (first or 
+	// second). Then we search the stack of first (or second) prism in Z
+	//
+	//
 
-	// Get user coordinates of cell containing point
+	// Check if point is in "first" triangle (0,0), (1,0), (1,1)
 	//
-	double coords0[3], coords1[3];
-	GetUserCoordinates(indices0, coords0);
-	GetUserCoordinates(indices1, coords1);
+	double lambda[3];
+	double pt[] = {coords[0],coords[1]};
+	Size_tArr3 iv = {indices[0], indices[0]+1, indices[0]+1};
+	Size_tArr3 jv = {indices[1], indices[1], indices[1]+1};
+	double tverts[] = {
+		_xcoords[iv[0]], _ycoords[jv[0]],
+		_xcoords[iv[1]], _ycoords[jv[1]],
+		_xcoords[iv[2]], _ycoords[jv[2]]
+	};
+	
+	bool inside = VAPoR::BarycentricCoordsTri(tverts, pt, lambda);
+	if (! inside) {
 
+		// Not in first triangle. 
+		// Now check if point is in "second" triangle (0,0), (1,1), (0,1)
+		//
+		iv = {indices[0], indices[0]+1, indices[0]};
+		jv = {indices[1], indices[1]+1, indices[1]+1};
+		double tverts[] = {
+			_xcoords[iv[0]], _ycoords[jv[0]],
+			_xcoords[iv[1]], _ycoords[jv[1]],
+			_xcoords[iv[2]], _ycoords[jv[2]]
+		};
+
+		inside = VAPoR::BarycentricCoordsTri(tverts, pt, lambda);
+
+		// Mathematically this shouldn't happen if _sg2d.GetIndicesCell() 
+		// returns true, but have to contend with floating point roundoff
+		//
+		if (! inside) return(false);
+	}
 
-	double x = coords[0];
-	double y = coords[1];
-	double z = coords[2];
-	double x0 = coords0[0];
-	double y0 = coords0[1];
-	double z0 = coords0[2];
-	double x1 = coords1[0];
-	double y1 = coords1[1];
-	double z1 = coords1[2];
+	float z0, z1;
 
+	// Check cached z values from last search first
 	//
-	// Calculate interpolation weights. We always interpolate along
-	// the varying dimension last (the kwgt)
-	//
-	z0 = _interpolateVaryingCoord(indices0[0],indices0[1],indices0[2],x,y);
-	z1 = _interpolateVaryingCoord(indices0[0],indices0[1],indices1[2],x,y);
+	if (((coords[2]-_insideGridCache.z0) * (coords[2]-_insideGridCache.z1)) < 0.0) {
+		indices[2] = _insideGridCache.k;
+		z0 = _insideGridCache.z0;
+		z1 = _insideGridCache.z1;
+	}
+	else {
+		 
+
+		// Find k index of cell containing z. Already know i and j indices
+		//
+		vector <double> zcoords;
+
+		size_t nz = GetDimensions()[2];
+		for (int kk=0; kk<nz; kk++) {
+
+			// Interpolate Z coordinate across triangle
+			//
+			float zk = 
+				_zrg.AccessIJK(iv[0], jv[0], kk) * lambda[0] +
+				_zrg.AccessIJK(iv[1], jv[1], kk) * lambda[1] +
+				_zrg.AccessIJK(iv[2], jv[2], kk) * lambda[2];
+
+				zcoords.push_back(zk);
+		}
+
+		if (! Wasp::BinarySearchRange(zcoords, coords[2], indices[2])) return(false);
```

Comment:
I believe it's better readability and maintenance if you call one of the std binary search functions. That is, std::lower_bound, std::upper_bound, or std::binary_search.

Reply:
Ha ha. I did this, but the implementation with lower_bound/upper_bound is actually more code than what I had. Oh well. I've committed it.

Reply:
Well, my understanding is that a user (e.g. a LayeredGrid) directly uses binary search functionality provided by std::lower_bound or std::upper_bound, but not use a wrapper (e.g. wasp::BinarySearchRange). Or in other words, if no new code uses wasp::BinarySearchRange() then that wrapper can go away after a few release cycles. Anyway, it's a minor issue.